### PR TITLE
Randomize Stream::merge to improve the throughput.

### PR DESCRIPTION
**Pull request by Razican**
This implements #490. Let me know of any improvements that can be done!

### CODE REVIEW

**Comment by Razican** [comment]
Hi, so I made 3 commits. The first one was a bit of a fail. Tests passed on my laoptop because this was an unstable function.

The second commit is what I think it should look like: one function to not duplicate code, but I was getting an error saying that `first`/`second` didn't implement `Stream`, which they do. So I don't understand the issue.

Duplicating the code in the two branches of the `if` works, though.

Do you know why the second commit might be failing?

**Comment by Razican** [comment]
Hi, I fixed the deduplication of the code, so I think this is ready for merging.

File: src/stream/stream/merge.rs [hunk 1]
```diff
@@ -40,14 +44,30 @@ where
...
+    first: Pin<&mut F>,
+    second: Pin<&mut S>,
+    cx: &mut Context<'_>,
+) -> Poll<Option<I>>
+where
+    F: Stream<Item = I>,
+    S: Stream<Item = I>,
+{
+    match first.poll_next(cx) {
+        Poll::Ready(Some(item)) => Poll::Ready(Some(item)),
```
Comment (k-nasa) [inline src/stream/stream/merge.rs]:
```suggestion
        Poll::Ready(item) => Poll::Ready(item),
```
I think either is fine, but I thought it was a little redundant.
However, the above changes will not match None, so the order must be changed.
Which do you think is better?

```rust
Poll::Ready(None) => second.poll_next(cx),
Poll::Ready(item) => Poll::Ready(item),
```

File: src/stream/stream/merge.rs [hunk 2]
```diff
@@ -40,14 +44,30 @@ where
...
+        Poll::Ready(None) => second.poll_next(cx),
+        Poll::Pending => match second.poll_next(cx) {
+            Poll::Ready(Some(item)) => Poll::Ready(Some(item)),
```
Comment (k-nasa) [inline src/stream/stream/merge.rs]:
```suggestion
            Poll::Ready(item) => Poll::Ready(item),
```
The same applies here

File: src/stream/stream/mod.rs [hunk 3]
```diff
@@ -1670,11 +1670,14 @@ extension_trait! {
...
-            assert_eq!(s.next().await, Some(1u8));
-            assert_eq!(s.next().await, Some(2u8));
-            assert_eq!(s.next().await, Some(3u8));
-            assert_eq!(s.next().await, None);
+            while let Some(n) = s.next().await {
+                lst.push(n)
+            }
+
+            lst.sort_unstable();
+            assert_eq!(&lst, &[1u8, 2u8, 3u8]);
```
Comment (k-nasa) [inline src/stream/stream/mod.rs]: 👍

File: src/stream/stream/mod.rs [hunk 4]
```diff
@@ -1670,11 +1670,14 @@ extension_trait! {
...
             let mut s = a.merge(b).merge(c);
+            let mut lst = Vec::new();
 
-            assert_eq!(s.next().await, Some(1u8));
-            assert_eq!(s.next().await, Some(2u8));
-            assert_eq!(s.next().await, Some(3u8));
-            assert_eq!(s.next().await, None);
+            while let Some(n) = s.next().await {
+                lst.push(n)
+            }
```
Comment (k-nasa) [inline src/stream/stream/mod.rs]: I think I can write like this.

```rust
use async_std::stream::FromStream;
let v = Vec::from_stream(stream).await;
```

Comment (Razican) [reply src/stream/stream/mod.rs]: Yes, this makes it more clear :)

File: src/stream/stream/merge.rs (hunk 1 above)
Comment (Razican) [reply src/stream/stream/merge.rs]: Hmmm yes, I will review this, better if we remove redundancy. I just used the code that was already in place, but this change looks better.

**Comment by Razican** [comment]
I made the requested changes and reduced the number of match arms in the second match (two arms had the same value).

**Pull request merged**